content/blog/2019-11-13-DynamicEdgeProfiling/index.md

```diff
@@ -0,0 +1,122 @@
++++
+title = "Dynamic Edge Profiling with Optimal Counter Placement"
+[extra]
+bio = """
+  [Katy](https://github.com/kavoor) is a Senior interested in Compilers and Systems Programming. She enjoys playing soccer and exploring Ithaca in her free time.
+  [Henry](https://www.linkedin.com/in/liuhenry4428)
+"""
++++
+
+## Introduction
+
+For this project, we wanted to implement an LLVM pass that dynamically profiled basic block edges. The naïve implementation would be to add a profiling mechanism to every edge in the program, but that comes with a significant amount of overhead. To minimize overhead, we strived to add profiling to the minimum number of edges necessary to obtain the same amount of profiling information. The key insight is that we can determine if an edge is traversed using profiling information from preceding and succeeding edges if the CFG structure ensures the program can only run through a simple path between those profiled edges.
+
+## Algorithm
+To select the edges to profile, we used Knuth's algorithm:
+
+---
+<!-- ![](https://i.imgur.com/NbavLnq.png) -->
+<img src="array.png" style="max-width: 100%" >
+---
+Essentially, this algorithm first finds the (acyclic) minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.
```

Comment:
To state the obvious, minimum spanning trees, because they are trees, are acyclic by definition—so that parenthetical may confuse more than it clarifies.